Refactor codegen file sharding

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62188
* #62187
* #62186
* #62278
* #62185
* __->__ #62184

File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

Differential Revision: [D29962050](https://our.internmc.facebook.com/intern/diff/D29962050)

peterbell10 committed:
Refactor codegen file sharding

File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

[ghstack-poisoned]

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

[ghstack-poisoned]

albanD commented on tools/codegen/gen.py:
```diff
@@ -837,9 +847,55 @@ def write_with_template(self, filename: str, template_fn: str,
+            key_fn: Callable[[T], str],
+            env_callable: Callable[[T], Dict[str, List[str]]],
+            num_shards: int,
+            base_env: Dict[str, Any] = {},
```
I think the empty dict here is actually wrong as flake8 pointed out

peterbell10 replied:
It's actually fine because I never modify `base_env`,  but have changed to `None` to appease the linter.

albanD commented on tools/codegen/gen.py:
```diff
@@ -837,9 +847,55 @@ def write_with_template(self, filename: str, template_fn: str,
+        shards: List[Dict[str, Any]] = [
+            {**base_env, 'shard_id': f'_{i}'} for i in range(num_shards)]
+
+        def merge_env(into: Dict[str, List[str]], from_: Dict[str, List[str]]):
```
why from_ and not from?

peterbell10 replied:
`from` is a keyword in python.

albanD replied:
Ho right...

albanD commented on tools/codegen/gen.py:
```diff
@@ -837,9 +847,55 @@ def write_with_template(self, filename: str, template_fn: str,
+                    assert isinstance(into[k], list)
+                    into[k] += v
+                else:
+                    into[k] = list(v)
```
isn't v already a list here?

peterbell10 replied:
That's what I thought as well. But if you don't copy the list here, then the same list will exist in two dicts and items will get re-appended into the same list.

albanD replied:
Ok!

albanD commented on tools/codegen/gen.py:
```diff
@@ -790,6 +794,12 @@ def compute_registration_declarations(f: NativeFunction, backend_indices: Dict[D
 
+
+# String hash that's stable across different executions, unlike builtin hash
+def string_stable_hash(s: str) -> int:
```
This is going to reshuffle all the functions that were previously in given files right?
Did you check that the file length are still balanced in practice?

peterbell10 replied:
No hash function is perfect, but I think a cryptographic has is going to be more reliable than adding all the ASCII characters together, like was done previously.

Here are the real-world distributions:

```
Sharding Operators.cpp, splits are:
 [427, 434, 459, 438, 381]
[341/1504] Generating ../../torch/csrc/autograd...esting/_internal/generated/annotated_fn_args.py
Sharding python_functions.cpp, splits are:
 [96, 101, 107, 118, 98]
Sharding python_torch_functions.cpp, splits are:
 [232, 267, 245]
Sharding VariableType.cpp, splits are:
 [251, 250, 283, 280, 237]
Sharding TraceType.cpp, splits are:
 [410, 422, 456, 461, 376]
```

albanD replied:
Looks good! Thanks for the numbers!

albanD reviewed (commented):
Nice refactoring! Thanks for doing that.
A couple small comments but looks good otherwise.

Do you know what is the exact impact on the generated code (apart from shuffling the functions around)?

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

[ghstack-poisoned]

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

[ghstack-poisoned]

albanD reviewed (approved):
LGTM!

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

[ghstack-poisoned]

albanD commented:
@peterbell10 could you check what's the runtime implication of this whole stack? By timing a standalone execution of the codegen. Just to make sure we're not increasing that by too much.

peterbell10 commented:
Using the following command:
```
time python -m tools.codegen.gen --source-path aten/src/ATen --install_dir build/aten/src/ATen
```

I get 5.616 s on master and 5.550 s, i.e. it's ~65 ms faster.

peterbell10 commented:
Using `setup_helpers/generate_code.py` I see an even bigger time saving:
```
time python tools/setup_helpers/generate_code.py --declarations-path "build/aten/src/ATen/Declarations.yaml" --native-functions-path "aten/src/ATen/native/native_functions.yaml" --nn-path "aten/src" --force_schema_registration
```

On master this takes 4.950 s and after this stack 3.385 s. For a saving of 1.6 seconds.

peterbell10 commented:
This saving is probably because for each native function, the old code generated everything three times.  Once for the shard, once for the everything file and once for the header (which didn't even use it). Whereas this only generates it once per function.

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

[ghstack-poisoned]

albanD commented:
Sounds great!
Thanks for the timings

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29962050).

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

Differential Revision: [D29962050](https://our.internmc.facebook.com/intern/diff/D29962050)

[ghstack-poisoned]

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

Differential Revision: [D29962050](https://our.internmc.facebook.com/intern/diff/D29962050)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29962050).

albanD commented:
Wow that breaks a lot of internal stuff!
in particular failing to find `type_derived_method_definitions` in template.py

```python
  File "tools/setup_helpers/generate_code.py", line 187, in <module>
  File "tools/setup_helpers/generate_code.py", line 172, in main
  File "tools/setup_helpers/generate_code.py", line 72, in generate_code
  File "tools/autograd/gen_autograd.py", line 64, in gen_autograd
  File "tools/autograd/gen_variable_type.py", line 378, in gen_variable_type
  File "tools/codegen/gen.py", line 896, in write_sharded
  File "tools/codegen/gen.py", line 843, in write_with_template
  File "tools/codegen/code_template.py", line 76, in substitute
  File "tools/codegen/code_template.py", line 64, in replace
  File "tools/codegen/code_template.py", line 46, in lookup
KeyError: 'type_derived_method_definitions'
```

Not sure why, I'll need to investigate...

peterbell10 commented:
> It is actually code that is modified here. Not sure why it only fails on some internal build though

Is it possible the build has no functions that get generated in `VariableType.cpp`? That's a case I hadn't considered.

peterbell10 committed:
Update on "Refactor codegen file sharding"


File sharding is currently implemented twice, once for VariableType and once for
TraceType. This refactors the implementation into `FileManager` and also changes
it so template substitution is only done once and shared between the sharded
file and the "Everything" file.

Differential Revision: [D29962050](https://our.internmc.facebook.com/intern/diff/D29962050)

[ghstack-poisoned]

peterbell10 commented on tools/codegen/gen.py:
```diff
@@ -859,23 +858,30 @@ def write_sharded(
+            for shard in all_shards:
                 shard.update(base_env)
 
+        for key in sharded_keys:
+            for shard in all_shards:
+                if key in shard:
+                    assert isinstance(shard[key], list), "sharded keys in base_env must be a list"
+                    shard[key] = shard[key].copy()
+                else:
+                    shard[key] = []
```
This loop ensures the key always exists in the environment, even if none of the functions hashed into that file.

albanD replied:
Awesome, thanks!

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29962050).
